miden-lib/asm/sat/prologue.masm

```diff
@@ -0,0 +1,363 @@
+#! - iah is the initial account hash of the account that the transaction is being executed against.
+#! - nc is the nullifier commitment of the transaction. This is a sequential hash of all
+#!   (nullifier, script_root) pairs for the notes consumed in the transaction.
+proc.process_public_inputs
+    # store the block hash
+    push.BLK_HASH_PTR mem_storew dropw
+
+    # store the account id
+    push.ACCT_ID_PTR mem_store
+
+    # store the initial account hash
+    push.ACCT_HASH_PTR mem_storew dropw
+
+    # store the nullifier commitment
+    push.NULLIFIER_COM_PTR mem_storew dropw
+end
+
+# ACCOUNT DATA
+# =================================================================================================
+
+#! Process the account data provided via the advice provider. This involves reading the data from
+#! the advice provider and storing it at the appropriate memory addresses. As the account data is
+#! read from the advice provider, the account hash is computed. It is asserted that the computed
+#! account hash matches the account hash provided via the stack public inputs. It is also asserted
+#! that the account id matches the account id provided via the stack public inputs.
+#!
+#! Stack: []
+#! Advice Tape: [aid, ZERO, ZERO, nonce, avr3, avr2, avr1, avr0, asr3, asr2, asr1, asr0, acr3,
+#!               acr2, acr1, acr0]
+#! Output: []
+#!
+#! - aid is the account id.
+#! - nonce is the account nonce.
+#! - avr is the account vault root.
+#! - asr is the account storage root.
+#! - acr is the account code root.
+proc.process_acct_data
+    # compute the address to store the account data
+    push.ACCT_DATA_PTR.1 add
+
+    # prepare the stack for absorbing account data
+    padw padw padw
+
+    # absorb the account data
+    adv_pipe adv_pipe
+
+    # extract digest from hasher rate elements (h_0, ..., h_3)
+    dropw swapw dropw
+
+    # assert that the account hash matches the hash provided by the stack public input
+    padw push.ACCT_HASH_PTR mem_loadw assert_eqw
+
+    # assert the account id matches the account id provided via stack public inputs
+    push.ACCT_ID_PTR mem_load
+    push.ACCT_DATA_PTR.1 add mem_load assert_eq
+
+    # clear the stack
+    drop
+end
+
+# CONSUMED NOTES DATA
+# =================================================================================================
+
+#! Computes a pointer to the memory address at which the data associated with a note with index i
+#! is stored.
+#!
+#! Stack: [i]
+#! Output: [ptr]
+proc.compute_note_data_ptr
+    # compute the address to store the consumed note data
+    push.1 add push.1024 mul push.CONSUMED_NOTE_DATA_OFFSET add
+end
+
+#! Computes a pointer to the memory address at which the nullifier associated a note with index i
+#! is stored.
+#!
+#! Stack: [i]
+#! Output: [ptr]
+proc.compute_nullifier_ptr
+    # compute the address to store the nullifier
+    push.CONSUMED_NOTE_DATA_OFFSET.1 add add
```

Comment:
Is this not the same as the location of serial number? Should we have a different offset for nullifiers and have a constant for it - something like `CONSUMED_NOTE_NULLIFIER_OFFSET`?

Reply:
Not quite.  All nullifiers are stored together in a sequential block of memory.  The memory block ranges from $c+1$ to $c+1024$ where $c$ is the `CONSUMED_NOTE_DATA_OFFEST`.  This is how it's described in the issue - https://github.com/0xPolygonMiden/miden-base/issues/14. Maybe we should introduce a constant  `CONSUMED_NOTE_NULLIFIERS_OFFSET` to mark the beginning of this sequential block.